File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.py
def solution(topping):
    tdict = {}
    for t in topping:
        if t not in tdict:
            tdict[t] = 1
        else:
            tdict[t] += 1
    
    adict = {}    
    answer = 0
    for i in topping:
        tdict[i] -= 1
        if i not in adict:
            adict[i] = 1
        else: 
            adict[i] += 1
            
        if tdict[i] == 0:
            tdict.pop(i)
        
        if len(tdict) == len(adict):
            answer += 1 
    return answer

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.cpp
#include <string>
#include <vector>
#include <unordered_map>
using namespace std;

int solution(vector<int> topping) {
    int answer = 0, cake;
    unordered_map <int,int> base, cmp;
    for(auto c : topping) {
        base[c]++;
    }
    cake = base.size(); // 케이크에 올라간 토핑의 총 개수 
    for(auto c : topping) {
        cmp[c]++;
        base[c]--;
        if (base[c] == 0){
            cake--;
        }
        if (cake == cmp.size()) {
            answer++;
        }
    }
    return answer;
}